File: src/main/lyrics.js
'use strict';

const LRCLIB_BASE = 'https://lrclib.net/api';
const USER_AGENT = 'lyric-overlay/0.1.0 (personal project)';

/**
 * Noise patterns found in media titles, especially browser/YouTube sessions.
 * Verified against a live SMTC sample:
 *   "'नalla' Freestyle (Visualizer) | Seedhe Maut x DJ SA"
 */
const TITLE_NOISE = [
  /\((?:official\s+)?(?:music\s+)?(?:video|audio|visualizer|lyrics?|lyric video|mv|hd|4k)\)/gi,
  /\[(?:official\s+)?(?:music\s+)?(?:video|audio|visualizer|lyrics?|lyric video|mv|hd|4k)\]/gi,
  /\((?:prod\.?|prod by)[^)]*\)/gi,
  /\b(?:official\s+)?(?:music\s+)?video\b/gi,
  /\bfull\s+(?:song|video|album)\b/gi,
  /\bremaster(?:ed)?(?:\s+\d{4})?\b/gi,
  // More browser/YouTube shapes seen in the wild.
  /\b(?:official\s+)?(?:audio|lyrical|lyric)\b/gi,
  /\bout\s+now\b/gi,
  /\s*[-–—]\s*youtube\s*$/gi,
];

/**
 * Words that mark a *different recording* of the same song. A studio LRC file
 * lines up with the studio cut and nothing else, so matching "Song (Live)" to
 * the studio entry yields lyrics that drift immediately. Duration alone does
 * not catch this — plenty of remixes run within a few seconds of the original.
 */
const VERSION_TOKENS = [
  'live', 'remix', 'acoustic', 'unplugged', 'instrumental', 'slowed', 'reverb',
  'sped', 'speedup', 'remastered', 'cover', 'karaoke', 'demo', 'extended',
  'mashup', 'reprise', 'bonus', 'edit', 'rework', 'flip', 'bootleg',
];

/**
 * Version markers present in a title, taken from the RAW string — cleanTitle
 * strips several of these, so this must run before it.
 *
 * @param {string} raw
 * @returns {Set<string>}
 */
function versionTags(raw) {
  const found = new Set();
  if (!raw) return found;
  const words = new Set(normalize(raw).split(' ').filter(Boolean));
  for (const token of VERSION_TOKENS) if (words.has(token)) found.add(token);
  return found;
}

/**
 * Strip platform noise from an artist credit.
 *
 * YouTube Music auto-generated channels report the artist as "Seedhe Maut -
 * Topic", and VEVO uploads append "VEVO". Both poison artist matching against
 * LRCLIB and the artwork APIs, and the "- Topic" suffix also leaks into the
 * on-screen artist label and the dancer registry lookup.
 *
 * @param {string} raw Raw artist string from SMTC.
 * @returns {string} Cleaned artist, or the original if cleaning emptied it.
 */
function cleanArtist(raw) {
  if (!raw) return '';
  const out = raw
    .replace(/\s*[-–—]\s*topic\s*$/i, '')
    // VEVO channel names are concatenated ("TheWeekndVEVO"), so there is no
    // word boundary to anchor on — strip it as a suffix instead.
    .replace(/vevo\s*$/i, '')
    .replace(/\s*[-–—]\s*official(?:\s+(?:channel|artist))?\s*$/i, '')
    .replace(/\s*\((?:official)?\s*\)\s*/gi, ' ')
    .replace(/\s+/g, ' ')
    .trim();
  return out || raw.trim();
}

/**
 * Strip platform noise from a raw media title.
 *
 * Handles the common browser/YouTube shapes: parenthetical tags, a trailing
 * "| Artist x Producer" credit block, and "feat./ft." suffixes.
 *
 * @param {string} raw Raw title from SMTC.
 * @returns {string} Cleaned title, or the original if cleaning emptied it.
 */
function cleanTitle(raw) {
  if (!raw) return '';
  let out = raw;

  // Everything after a pipe on a YouTube-style title is credits, not the song.
  const pipeIndex = out.indexOf('|');
  if (pipeIndex > 0) out = out.slice(0, pipeIndex);

  for (const pattern of TITLE_NOISE) out = out.replace(pattern, ' ');

  out = out
    .replace(/\((?:feat\.?|ft\.?|with)[^)]*\)/gi, ' ')
    .replace(/\b(?:feat\.?|ft\.?)\s+.*$/gi, ' ')
    .replace(/["'`‘’“”]/g, ' ')
    .replace(/\s*[-–—]\s*$/, '')
    .replace(/\s+/g, ' ')
    .trim();

  return out || raw.trim();
}

/**
 * Normalize a string for comparison: lowercase, strip Latin diacritics and
 * punctuation, collapse whitespace. Non-Latin scripts (e.g. Devanagari) are
 * preserved rather than stripped, since they carry real title content.
 *
 * @param {string} value
 * @returns {string}
 */
function normalize(value) {
  if (!value) return '';
  return value
    .normalize('NFKD')
    .replace(/[̀-ͯ]/g, '')
    .toLowerCase()
    .replace(/[^\p{L}\p{N}\s]/gu, ' ')
    .replace(/\s+/g, ' ')
    .trim();
}

/**
 * Token-overlap similarity in [0, 1].
 * @param {string} a
 * @param {string} b
 * @returns {number}
 */
function tokenSimilarity(a, b) {
  const setA = new Set(normalize(a).split(' ').filter(Boolean));
  const setB = new Set(normalize(b).split(' ').filter(Boolean));
  if (setA.size === 0 || setB.size === 0) return 0;

  let shared = 0;
  for (const token of setA) if (setB.has(token)) shared += 1;
  return shared / Math.max(setA.size, setB.size);
}

/**
 * Score one LRCLIB candidate against the track we're looking for.
 *
 * LRCLIB entries are themselves messy — verified examples include
 * "11K - Seedhe Maut" and "Hausla - Seedhe Maut", where the artist is baked
 * into trackName. So both sides get compared as a combined haystack too.
 *
 * @param {object} candidate LRCLIB search result.
 * @param {{title: string, artist: string, durationMs: number}} target
 * @returns {number} Score; higher is better. Negative means reject.
 */
function scoreCandidate(candidate, target) {
  if (!candidate.syncedLyrics) return -1; // Line-level sync is the whole point.

  const cleanedTarget = cleanTitle(target.title);

  const titleScore = Math.max(
    tokenSimilarity(cleanedTarget, candidate.trackName || ''),
    // Fallback: compare against the candidate's combined track+artist string,
    // which catches entries with the artist embedded in the title.
    tokenSimilarity(cleanedTarget, `${candidate.trackName} ${candidate.artistName}`)
  );

  const cleanedArtist = cleanArtist(target.artist);
  const artistScore = cleanedArtist
    ? tokenSimilarity(cleanedArtist, candidate.artistName || '')
    : 0;

  let score = titleScore * 2 + artistScore;

  // Duration proximity is the strongest disambiguator between remixes,
  // live cuts and the studio version. LRCLIB reports seconds.
  if (target.durationMs > 0 && candidate.duration > 0) {
    const deltaSec = Math.abs(candidate.duration - target.durationMs / 1000);
    if (deltaSec <= 2) score += 1.5;
    else if (deltaSec <= 5) score += 0.75;
    else if (deltaSec > 20) score -= 1.0;
  }

  /*
    Version guard. Duration catches gross mismatches, but a remix or a slowed
    edit often lands within a few seconds of the original, scores well on
    title/artist, and then plays with lyrics that drift from the first line.
    Penalise each version marker present on one side and missing on the other.
  */
  const wantVersions = versionTags(target.title);
  const haveVersions = versionTags(`${candidate.trackName || ''} ${candidate.albumName || ''}`);
  let versionMismatch = 0;
  for (const tag of wantVersions) if (!haveVersions.has(tag)) versionMismatch += 1;
  for (const tag of haveVersions) if (!wantVersions.has(tag)) versionMismatch += 1;
  score -= versionMismatch * 1.25;

  return score;
}

/**
 * Parse an LRC document into ordered, timestamped cues.
 *
 * @param {string} lrc Raw LRC text.
 * @returns {Array<{timeMs: number, text: string}>} Sorted cues.
 */
function parseLrc(lrc) {
  const cues = [];
  const lineTag = /\[(\d{1,3}):(\d{2})(?:[.:](\d{1,3}))?\]/g;

  for (const rawLine of lrc.split(/\r?\n/)) {
    lineTag.lastIndex = 0;
    const stamps = [];
    let match;
    while ((match = lineTag.exec(rawLine)) !== null) {
      const minutes = Number(match[1]);
      const seconds = Number(match[2]);
      const fracRaw = match[3] ?? '0';
      // Normalize 1-3 digit fractions to milliseconds.
      const fraction = Number(fracRaw.padEnd(3, '0').slice(0, 3));
      stamps.push(minutes * 60_000 + seconds * 1000 + fraction);
    }
    if (stamps.length === 0) continue;

    const text = rawLine.replace(lineTag, '').trim();
    for (const timeMs of stamps) cues.push({ timeMs, text });
  }

  cues.sort((a, b) => a.timeMs - b.timeMs);
  return cues;
}

/**
 * Fraction of script letters that belong to the Devanagari script.
 *
 * Both numerator and denominator count only letters (`\p{L}`), so Devanagari
 * combining vowel marks (matras, category Mn) don't skew the result — otherwise
 * a fully-Devanagari string can exceed a ratio of 1.
 *
 * @param {string} text
 * @returns {number} 0..1
 */
function devanagariRatio(text) {
  if (!text) return 0;
  const letters = text.match(/\p{L}/gu) || [];
  if (letters.length === 0) return 0;
  const deva = letters.filter((ch) => /\p{Script=Devanagari}/u.test(ch));
  return deva.length / letters.length;
}

/**
 * Whether a parsed cue list is predominantly Devanagari script.
 * @param {Array<{text: string}>} cues
 * @returns {boolean}
 */
function isDevanagariCues(cues) {
  const sample = cues
    .slice(0, 12)
    .map((c) => c.text)
    .join(' ');
  return devanagariRatio(sample) >= 0.4;
}

/**
 * Fraction of script letters that belong to the Gurmukhi (Punjabi) script.
 * @param {string} text
 * @returns {number} 0..1
 */
function gurmukhiRatio(text) {
  if (!text) return 0;
  const letters = text.match(/\p{L}/gu) || [];
  if (letters.length === 0) return 0;
  const guru = letters.filter((ch) => /\p{Script=Gurmukhi}/u.test(ch));
  return guru.length / letters.length;
}

/**
 * Common romanized Hindi/Punjabi function words. Their presence in Latin-script
 * lyrics is a strong signal the language is Indic rather than English, which
 * lets us auto-offer English translation without an API round-trip just to
 * detect language.
 */
const INDIC_ROMAN_TOKENS = new Set([
  'hai', 'nahi', 'nai', 'ni', 'mera', 'meri', 'tera', 'teri', 'tu', 'tum', 'main',
  'mai', 'kya', 'kyu', 'kyun', 'ke', 'ki', 'ka', 'ko', 'se', 'me', 'mein', 'pe',
  'par', 'aur', 'ab', 'jo', 'wo', 'woh', 'yeh', 'ye', 'hoke', 'hona', 'karke',
  'gaya', 'raha', 'rahe', 'kar', 'karo', 'chal', 'dil', 'pyaar', 'yaar', 'bhai',
  'apna', 'apne', 'saara', 'sab', 'kuch', 'bina', 'zindagi', 'sunke', 'rona',
  'gaane', 'itna', 'aata', 'jaata', 'wala', 'wale', 'haan', 'bol', 'bole',
]);

/**
 * Heuristic: do these cues look like Hindi/Punjabi (native script or romanized)?
 *
 * Used to decide whether to auto-offer an English translation line.
 *
 * @param {Array<{text: string}>} cues
 * @returns {{indic: boolean, script: 'devanagari'|'gurmukhi'|'latin'}}
 */
function detectIndic(cues) {
  const sample = cues.slice(0, 20).map((c) => c.text).join(' ');

  if (devanagariRatio(sample) >= 0.4) return { indic: true, script: 'devanagari' };
  if (gurmukhiRatio(sample) >= 0.4) return { indic: true, script: 'gurmukhi' };

  const words = normalize(sample).split(' ').filter(Boolean);
  if (words.length === 0) return { indic: false, script: 'latin' };

  let hits = 0;
  for (const word of words) if (INDIC_ROMAN_TOKENS.has(word)) hits += 1;

  // A few function-word hits per line's worth of text is a reliable signal.
  const ratio = hits / words.length;
  return { indic: ratio >= 0.06 || hits >= 4, script: 'latin' };
}

/**
 * Look up synced lyrics for a track.
 *
 * Uses LRCLIB's free-text `q=` endpoint. The structured `track_name`/
 * `artist_name` endpoint was verified to be too strict — it returned zero
 * results for tracks that free-text search matched successfully.
 *
 * Scans all candidates for a Devanagari-native synced entry as well as the
 * best overall match, so a native Hindi-script version is used when it exists
 * rather than falling back to machine transliteration.
 *
 * @param {{title: string, artist: string, durationMs: number}} track
 * @returns {Promise<{cues: Array, cuesDevanagari: Array|null, source: object}|null>}
 */
async function fetchSyncedLyrics(track) {
  const query = [cleanTitle(track.title), track.artist].filter(Boolean).join(' ');
  if (!query.trim()) return null;

  const url = `${LRCLIB_BASE}/search?q=${encodeURIComponent(query)}`;

  let results;
  try {
    const response = await fetch(url, {
      headers: { 'User-Agent': USER_AGENT },
      signal: AbortSignal.timeout(15_000),
    });
    if (!response.ok) {
      throw new Error(`LRCLIB responded ${response.status}`);
    }
    results = await response.json();
  } catch (err) {
    throw new Error(`Lyric lookup failed: ${err.message}`);
  }

  if (!Array.isArray(results) || results.length === 0) return null;

  let best = null;
  let bestScore = 0;
  let devaBest = null;
  let devaBestScore = 0;

  for (const candidate of results) {
    const score = scoreCandidate(candidate, track);
    if (score > bestScore) {
      bestScore = score;
      best = candidate;
    }
    // Track the best Devanagari-script candidate separately. Use a lower
    // acceptance bar for it (0.5) — a native-script version that matches the
    // song reasonably well is worth surfacing even if its metadata is noisier.
    if (score > devaBestScore && candidate.syncedLyrics && devanagariRatio(candidate.syncedLyrics) >= 0.4) {
      devaBestScore = score;
      devaBest = candidate;
    }
  }

  // Require a meaningful match rather than returning the least-bad result.
  if (!best || bestScore < 1.0) return null;

  const cues = parseLrc(best.syncedLyrics);
  if (cues.length === 0) return null;

  // If the primary match is itself Devanagari, it doubles as the native track.
  let cuesDevanagari = null;
  if (isDevanagariCues(cues)) {
    cuesDevanagari = cues;
  } else if (devaBest && devaBestScore >= 0.5) {
    const parsed = parseLrc(devaBest.syncedLyrics);
    if (parsed.length > 0) cuesDevanagari = parsed;
  }

  return {
    cues,
    cuesDevanagari,
    source: {
      id: best.id,
      trackName: best.trackName,
      artistName: best.artistName,
      duration: best.duration,
      score: Number(bestScore.toFixed(2)),
    },
  };
}

module.exports = {
  cleanTitle,
  cleanArtist,
  versionTags,
  normalize,
  tokenSimilarity,
  scoreCandidate,
  parseLrc,
  devanagariRatio,
  gurmukhiRatio,
  isDevanagariCues,
  detectIndic,
  fetchSyncedLyrics,
};

File: src-tauri/src/lyrics.rs
//! Synced-lyric lookup against LRCLIB.
//!
//! A focused Rust port of the parts of the old `src/main/lyrics.js` that the
//! core loop needs: clean the messy SMTC title/artist, search LRCLIB's
//! free-text endpoint, score candidates by title/artist/duration/version, and
//! parse the winning LRC into timestamped cues (folding textless gap markers
//! into `endMs`). NetEase/Kugou/plain-lyric fallbacks and the LLM passes are
//! later phases.

use std::sync::LazyLock;

use regex::Regex;
use serde::Serialize;
use serde_json::Value;

const LRCLIB_SEARCH: &str = "https://lrclib.net/api/search";
const USER_AGENT: &str = "lyric-overlay/0.25.0 (personal project)";

/// The track we are looking up.
pub struct Track {
    pub title: String,
    pub artist: String,
    pub duration_ms: i64,
}

/// One timestamped lyric line. Serialises to the shape the renderer expects.
#[derive(Serialize, Clone)]
pub struct Cue {
    #[serde(rename = "timeMs")]
    pub time_ms: i64,
    pub text: String,
    #[serde(rename = "endMs", skip_serializing_if = "Option::is_none")]
    pub end_ms: Option<i64>,
}

/// Words that mark a *different recording* — a studio LRC won't line up with a
/// live/remix cut even at a similar duration.
const VERSION_TOKENS: &[&str] = &[
    "live", "remix", "acoustic", "unplugged", "instrumental", "slowed", "reverb",
    "sped", "speedup", "remastered", "cover", "karaoke", "demo", "extended",
    "mashup", "reprise", "bonus", "edit", "rework", "flip", "bootleg",
];

static NOISE_PATTERNS: LazyLock<Vec<Regex>> = LazyLock::new(|| {
    [
        r"(?i)\((?:official\s+)?(?:music\s+)?(?:video|audio|visualizer|lyrics?|lyric video|mv|hd|4k)\)",
        r"(?i)\[(?:official\s+)?(?:music\s+)?(?:video|audio|visualizer|lyrics?|lyric video|mv|hd|4k)\]",
        r"(?i)\((?:prod\.?|prod by)[^)]*\)",
        r"(?i)\b(?:official\s+)?(?:music\s+)?video\b",
        r"(?i)\bfull\s+(?:song|video|album)\b",
        r"(?i)\bremaster(?:ed)?(?:\s+\d{4})?\b",
        r"(?i)\b(?:official\s+)?(?:audio|lyrical|lyric)\b",
        r"(?i)\bout\s+now\b",
    ]
    .iter()
    .map(|p| Regex::new(p).unwrap())
    .collect()
});

static FEAT_PARENS: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"(?i)\((?:feat\.?|ft\.?|with)[^)]*\)").unwrap());
static FEAT_TAIL: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"(?i)\b(?:feat\.?|ft\.?)\s+.*$").unwrap());
static NON_ALNUM: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"[^\p{L}\p{N}\s]").unwrap());
static WS: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\s+").unwrap());
static LINE_TAG: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"\[(\d{1,3}):(\d{2})(?:[.:](\d{1,3}))?\]").unwrap());

/// Lowercase, strip punctuation, collapse whitespace. Keeps non-Latin scripts.
fn normalize(value: &str) -> String {
    let lowered = value.to_lowercase();
    let no_punct = NON_ALNUM.replace_all(&lowered, " ");
    WS.replace_all(&no_punct, " ").trim().to_string()
}

/// Token-overlap similarity in [0, 1].
pub fn token_similarity(a: &str, b: &str) -> f64 {
    let na = normalize(a);
    let nb = normalize(b);
    let sa: std::collections::HashSet<&str> = na.split(' ').filter(|s| !s.is_empty()).collect();
    let sb: std::collections::HashSet<&str> = nb.split(' ').filter(|s| !s.is_empty()).collect();
    if sa.is_empty() || sb.is_empty() {
        return 0.0;
    }
    let shared = sa.iter().filter(|t| sb.contains(*t)).count();
    shared as f64 / sa.len().max(sb.len()) as f64
}

/// Version markers present in a raw title/album string.
pub fn version_tags(raw: &str) -> std::collections::HashSet<String> {
    let words: std::collections::HashSet<String> =
        normalize(raw).split(' ').filter(|s| !s.is_empty()).map(|s| s.to_string()).collect();
    VERSION_TOKENS
        .iter()
        .filter(|t| words.contains(**t))
        .map(|t| t.to_string())
        .collect()
}

/// Strip platform noise from a raw media title.
pub fn clean_title(raw: &str) -> String {
    let mut out = raw.to_string();
    if let Some(idx) = out.find('|') {
        out = out[..idx].to_string();
    }
    for re in NOISE_PATTERNS.iter() {
        out = re.replace_all(&out, " ").to_string();
    }
    out = FEAT_PARENS.replace_all(&out, " ").to_string();
    out = FEAT_TAIL.replace_all(&out, " ").to_string();
    out = out.replace(['"', '\'', '`', '‘', '’', '“', '”'], " ");
    let out = WS.replace_all(&out, " ").trim().to_string();
    if out.is_empty() {
        raw.trim().to_string()
    } else {
        out
    }
}

/// Strip "- Topic" / VEVO from an artist credit.
pub fn clean_artist(raw: &str) -> String {
    let lowered = raw.to_lowercase();
    let mut out = raw.to_string();
    if let Some(pos) = lowered.rfind("- topic") {
        out = out[..pos].to_string();
    }
    let trimmed = out.trim();
    let stripped = trimmed.strip_suffix("VEVO").or_else(|| trimmed.strip_suffix("vevo")).unwrap_or(trimmed);
    let cleaned = WS.replace_all(stripped, " ").trim().to_string();
    if cleaned.is_empty() {
        raw.trim().to_string()
    } else {
        cleaned
    }
}

/// Score one LRCLIB candidate. Requires syncedLyrics. Negative = reject.
fn score_candidate(candidate: &Value, track: &Track) -> f64 {
    if candidate.get("syncedLyrics").and_then(|v| v.as_str()).unwrap_or("").is_empty() {
        return -1.0;
    }
    let cleaned_title = clean_title(&track.title);
    let track_name = candidate.get("trackName").and_then(|v| v.as_str()).unwrap_or("");
    let artist_name = candidate.get("artistName").and_then(|v| v.as_str()).unwrap_or("");

    let title_score = token_similarity(&cleaned_title, track_name)
        .max(token_similarity(&cleaned_title, &format!("{track_name} {artist_name}")));

    let cleaned_artist = clean_artist(&track.artist);
    let artist_score = if cleaned_artist.is_empty() {
        0.0
    } else {
        token_similarity(&cleaned_artist, artist_name)
    };

    let mut score = title_score * 2.0 + artist_score;

    let cand_dur = candidate.get("duration").and_then(|v| v.as_f64()).unwrap_or(0.0);
    if track.duration_ms > 0 && cand_dur > 0.0 {
        let delta = (cand_dur - track.duration_ms as f64 / 1000.0).abs();
        if delta <= 2.0 {
            score += 1.5;
        } else if delta <= 5.0 {
            score += 0.75;
        } else if delta > 20.0 {
            score -= 1.0;
        }
    }

    let want = version_tags(&track.title);
    let album = candidate.get("albumName").and_then(|v| v.as_str()).unwrap_or("");
    let have = version_tags(&format!("{track_name} {album}"));
    let mut mismatch = 0;
    for tag in &want {
        if !have.contains(tag) {
            mismatch += 1;
        }
    }
    for tag in &have {
        if !want.contains(tag) {
            mismatch += 1;
        }
    }
    score - mismatch as f64 * 1.25
}

/// Parse an LRC document into ordered cues, folding textless stamps into `endMs`.
pub fn parse_lrc(lrc: &str) -> Vec<Cue> {
    let mut stamped: Vec<(i64, String)> = Vec::new();
    for raw_line in lrc.split('\n') {
        let mut stamps: Vec<i64> = Vec::new();
        for cap in LINE_TAG.captures_iter(raw_line) {
            let minutes: i64 = cap[1].parse().unwrap_or(0);
            let seconds: i64 = cap[2].parse().unwrap_or(0);
            let frac_raw = cap.get(3).map(|m| m.as_str()).unwrap_or("0");
            let mut frac = frac_raw.to_string();
            while frac.len() < 3 {
                frac.push('0');
            }
            let fraction: i64 = frac[..3].parse().unwrap_or(0);
            stamps.push(minutes * 60_000 + seconds * 1000 + fraction);
        }
        if stamps.is_empty() {
            continue;
        }
        let text = LINE_TAG.replace_all(raw_line, "").trim().to_string();
        for t in stamps {
            stamped.push((t, text.clone()));
        }
    }
    stamped.sort_by_key(|(t, _)| *t);

    // Fold textless markers into endMs on the line they close.
    let mut out: Vec<Cue> = Vec::new();
    for (time_ms, text) in stamped {
        if text.is_empty() {
            if let Some(last) = out.last_mut() {
                if last.end_ms.is_none() && time_ms > last.time_ms {
                    last.end_ms = Some(time_ms);
                }
            }
            continue;
        }
        out.push(Cue { time_ms, text, end_ms: None });
    }
    out
}

/// Look up synced lyrics for a track. Returns (cues, source label) or None.
pub fn fetch_synced(track: &Track) -> Option<(Vec<Cue>, Value)> {
    let query = format!("{} {}", clean_title(&track.title), track.artist);
    let query = query.trim();
    if query.is_empty() {
        return None;
    }

    let resp = ureq::get(LRCLIB_SEARCH)
        .set("User-Agent", USER_AGENT)
        .query("q", query)
        .timeout(std::time::Duration::from_secs(15))
        .call()
        .ok()?;
    let results: Value = resp.into_json().ok()?;
    let arr = results.as_array()?;
    if arr.is_empty() {
        return None;
    }

    let mut best: Option<&Value> = None;
    let mut best_score = 0.0_f64;
    for candidate in arr {
        let s = score_candidate(candidate, track);
        if s > best_score {
            best_score = s;
            best = Some(candidate);
        }
    }
    // Require a meaningful match, mirroring the JS floor.
    let best = best?;
    if best_score < 1.0 {
        return None;
    }
    let synced = best.get("syncedLyrics").and_then(|v| v.as_str())?;
    let cues = parse_lrc(synced);
    if cues.is_empty() {
        return None;
    }

    let source = serde_json::json!({
        "id": best.get("id"),
        "trackName": best.get("trackName"),
        "artistName": best.get("artistName"),
        "duration": best.get("duration"),
        "score": (best_score * 100.0).round() / 100.0,
    });
    Some((cues, source))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn clean_title_strips_noise_and_pipe_credits() {
        assert_eq!(clean_title("Blinding Lights (Official Video)"), "Blinding Lights");
        assert_eq!(clean_title("Song | Artist x Producer"), "Song");
        assert_eq!(clean_title("Track (feat. Someone)"), "Track");
        // Cleaning to empty falls back to the raw trimmed title.
        assert_eq!(clean_title("(Official Audio)"), "(Official Audio)");
    }

    #[test]
    fn clean_artist_strips_topic_and_vevo() {
        assert_eq!(clean_artist("Seedhe Maut - Topic"), "Seedhe Maut");
        assert_eq!(clean_artist("TheWeekndVEVO"), "TheWeeknd");
    }

    #[test]
    fn parse_lrc_folds_textless_markers_into_end_ms() {
        let lrc = "[00:14.15] Yeah\n[00:17.73]\n[00:27.85] I've been tryna call";
        let cues = parse_lrc(lrc);
        assert_eq!(cues.len(), 2);
        assert_eq!(cues[0].text, "Yeah");
        assert_eq!(cues[0].time_ms, 14_150);
        assert_eq!(cues[0].end_ms, Some(17_730)); // the blank line closed it
        assert_eq!(cues[1].text, "I've been tryna call");
        assert_eq!(cues[1].time_ms, 27_850);
    }

    #[test]
    fn score_rejects_candidates_without_synced_lyrics() {
        let track = Track { title: "X".into(), artist: "Y".into(), duration_ms: 0 };
        let cand = serde_json::json!({ "trackName": "X", "artistName": "Y" });
        assert!(score_candidate(&cand, &track) < 0.0);
    }

    // Live network test — run explicitly with `cargo test -- --ignored`.
    #[test]
    #[ignore]
    fn live_fetch_known_song() {
        let track = Track {
            title: "Blinding Lights".into(),
            artist: "The Weeknd".into(),
            duration_ms: 200_000,
        };
        let result = fetch_synced(&track);
        assert!(result.is_some(), "expected LRCLIB to have Blinding Lights");
        let (cues, source) = result.unwrap();
        assert!(cues.len() > 5, "expected multiple synced lines, got {}", cues.len());
        eprintln!("matched: {source}");
        eprintln!("first line: [{}ms] {}", cues[0].time_ms, cues[0].text);
    }
}
